### PR TITLE
[lsp-dart] Add outline and flutterOutline options

### PR DESCRIPTION
**Pull request by ericdallo**
image

### CODE REVIEW

File: lsp-dart.el [hunk 1]
```diff
@@ -80,6 +80,154 @@ Defaults to 0.9"
...
+          (window (display-buffer-in-side-window buffer lsp-dart-outline-position-params)))
+    (when (null focus?)
+      (select-window window)
+      (set-window-dedicated-p window t))))
+
+(defun lsp-dart--show-flutter-outline (focus?)
+  "Shows an Flutter outline tree.
+Focus on it if FOCUS? is non-null."
+  (-let* (((&hash "uri" "outline" (&hash "children")) (lsp-workspace-get-metadata "current-flutter-outline"
+                                                                                  (first (lsp-workspaces))))
```
Comment (yyoncho) [inline lsp-dart.el:224]: `(first (lsp-workspaces))` - `lsp-find-workspace`

File: lsp-dart.el [hunk 2]
```diff
@@ -120,10 +284,26 @@ PARAMS closing labels notification data sent from WORKSPACE."
...
+                    (closingLabels . ,lsp-dart-closing-labels)
+                    (outline . ,lsp-dart-outline)
+                    (flutterOutline . ,lsp-dart-flutter-outline))
+                  :notification-handlers (ht ("dart/textDocument/publishClosingLabels" 'lsp-dart--handle-closing-labels)
+                                             ("dart/textDocument/publishOutline" 'lsp-dart--handle-outline)
+                                             ("dart/textDocument/publishFlutterOutline" 'lsp-dart--handle-flutter-outline))
                   :server-id 'dart_analysis_server))
 
+;;;###autoload
+(with-eval-after-load 'lsp-treemacs
```
Comment (yyoncho) [inline lsp-dart.el:296]: all pieces that refer to lsp-treemacs should be wrapped in `(with-eval-after-load` . Without this, the travis will fail.

File: lsp-dart.el [hunk 3]
```diff
@@ -120,10 +284,26 @@ PARAMS closing labels notification data sent from WORKSPACE."
...
-                  :notification-handlers (ht ("dart/textDocument/publishClosingLabels" 'lsp-dart--handle-closing-labels))
+                    (closingLabels . ,lsp-dart-closing-labels)
+                    (outline . ,lsp-dart-outline)
+                    (flutterOutline . ,lsp-dart-flutter-outline))
+                  :notification-handlers (ht ("dart/textDocument/publishClosingLabels" 'lsp-dart--handle-closing-labels)
+                                             ("dart/textDocument/publishOutline" 'lsp-dart--handle-outline)
+                                             ("dart/textDocument/publishFlutterOutline" 'lsp-dart--handle-flutter-outline))
                   :server-id 'dart_analysis_server))
 
+;;;###autoload
```
Comment (yyoncho) [inline lsp-dart.el:295]: I don't think this autoload is needed.

File: lsp-dart.el [hunk 4]
```diff
@@ -80,6 +80,154 @@ Defaults to 0.9"
...
+   "*Flutter Outline*"))
+
+(defun lsp-dart--show-outline (focus?)
+  "Shows an outline tree.
+Focus on it if FOCUS? is non-null."
+  (-let* (((&hash "uri" "outline" (&hash "children")) (lsp-workspace-get-metadata "current-outline"
+                                                                                  (first (lsp-workspaces))))
+          (buffer (lsp-dart--build-outline-tree uri children))
+          (window (display-buffer-in-side-window buffer lsp-dart-outline-position-params)))
+    (when (null focus?)
```
Comment (yyoncho) [inline lsp-dart.el:216]: this `null` seems to be redundant.

File: lsp-dart.el [hunk 5]
```diff
@@ -80,6 +80,154 @@ Defaults to 0.9"
...
+  (seq-map (-lambda ((&hash "children"
+                            "element" (&hash "kind" "name" "parameters" "range")))
+             (let ((label (concat name (when parameters
+                                         (propertize (concat " " parameters)
+                                                     'face 'lsp-lens-face)))))
+               (list :key label
+                     :label label
+                     :icon (lsp-dart--outline-kind->icon kind)
+                     :children (unless (seq-empty-p children)
+                                 (lsp-dart--outline->tree uri children))
```
Comment (yyoncho) [inline lsp-dart.el:169]: in general, you can pass a function to :children to make the conversion lazy. You may skip this one if you do not expect this function to be called often with big data.

File: lsp-dart.el [hunk 6]
```diff
@@ -80,6 +80,154 @@ Defaults to 0.9"
...
+    (slot . 2)
+    (window-width . ,treemacs-width))
+  "The outline tree position params.
+Defaults to side following treemacs default."
+  :type 'list
+  :group 'lsp-dart
+  :package-version '(lsp-mode . "6.3"))
+
+(defcustom lsp-dart-flutter-outline-position-params
+  `((side . ,treemacs-position)
```
Comment (yyoncho) [inline lsp-dart.el:108]: you cannot use treemacs-position here like that - since we do not have a dependency on it.

Comment (ericdallo) [reply lsp-dart.el:108]: Wrapped into a `with-eval-after-load`

File: lsp-dart.el (hunk 2 above)
Comment (ericdallo) [reply lsp-dart.el:296]: Done, I don't know if I need to wrap the "private" functions given that the public one is already wrapped with `with-eval-after-load`

Comment (yyoncho) [reply lsp-dart.el:296]: everything that is undefined will cause test failures no matter if it is public or private.

**Comment by yyoncho** [comment]
Thank you!

**Comment by ericdallo** [comment]
Thank you for all the help @yyoncho !

**Pull request merged**